kolibri/auth/permissions.py

```diff
@@ -0,0 +1,349 @@
+
+    def user_can_update_object(self, user, obj):
+
+        # if no explicit `update` permissions were specified, defer to the `write` permissions
+        if "update" not in self.permissions_by_role:
+            return self._user_can_write_object(user, obj)
+
+        return self._check_role_permission(user, obj, "update")
+
+    def user_can_delete_object(self, user, obj):
+
+        # if no explicit `delete` permissions were specified, defer to the `write` permissions
+        if "delete" not in self.permissions_by_role:
+            return self._user_can_write_object(user, obj)
+
+        return self._check_role_permission(user, obj, "delete")
+
+    def _user_can_write_object(self, user, obj):
+
+        return self._check_role_permission(user, obj, "write")
+
+
+#####################################################################################################################
+### This section consists of some example permissions classes; later, these would likely live within other apps.  ###
+#####################################################################################################################
+
+class IsDeviceOwner(BasePermissions):
+    """
+    Permissions class that only allows access to a DeviceOwner, and denies access to all others.
+    """
+
+    def user_can_read_object(self, user, obj):
+        return user._is_device_owner
+
+    def _user_can_write_object(self, user, obj):
+        return user._is_device_owner
+
+
+class UserLogPermissions(RoleBasedPermissions):
+    """
+    A general Permissions class intended for log data pertaining to a particular user (such as records
+    of interactions with a piece of content). We allow the associated user herself, a coach for the user,
+    or an admin for the user to read the log, but only the user herself or an admin to modify it.
+    """
+
+    permissions_by_role = {
+        "read": ["self", "coach", "admin"],
+        "write": ["self", "admin"],
+    }
+
+    target = {
+        "type": "user",
+        "field": "user",
+    }
+
+
+class FacilityUserPermissions(RoleBasedPermissions):
+    """
+    A Permissions class intended for controlling access to the FacilityUser object itself.
+    """
+
+    permissions_by_role = {
+        "create": ["all"], # in terms of hard permissions, anyone can sign up, though we may limit this in interface
+        "read": ["self", "coach", "admin"],
+        "update": ["self", "admin"],
+        "delete": ["admin"],
+    }
+
+    target = {
+        "type": "user",
+        "field": ".",
+    }
+
+
+class CollectionPermissions(RoleBasedPermissions):
+    """
+    A Permissions class intended for controlling access to a Collection object; writing is only allowed
+    by an admin for the collection.
+    """
+
+    permissions_by_role = {
+        "read": ["all"],
+        "write": ["admin"],
+    }
+
+    target = {
+        "type": "collection",
+        "field": ".",
```

Comment:
Okay, so the interpretation of `target` is that if I assign this Permissions class to a Model, then the model must possess a field with the name indicated here, *and* that field must be of the "type" indicated. And it doesn't imply introspection using the builtin `type` function.

Is it possible to dispense with specifying `type` using introspection? For example

```python
if isinstance(obj, BaseUser):
   type = 'user'
elif isinstance(obj, Collection):
   type = 'collection'